linalg/ode.hpp

```diff
@@ -12,13 +12,91 @@
 #ifndef MFEM_ODE
 #define MFEM_ODE
 
+#include "../general/communication.hpp"
 #include "../config/config.hpp"
 #include "operator.hpp"
-#include "../general/communication.hpp"
+#include <vector>
+#include <memory>
 
 namespace mfem
 {
 
+/// An interface for storing state of previous timesteps
+class ODEStateData
+{
+public:
+   /// Get the maximum number of stored stages
+   virtual int MaxSize() const = 0;
+
+   /// Get the current number of stored stages
+   virtual int Size() const = 0;
+
+   /// Get the ith state vector
+   virtual void Get(int i, Vector &state) const = 0;
```

Comment:
Do we need this method? Isn't it equivalent to `state = Get(i)`?

Reply:
The current layout of classes `ODEStateData` and `ODEStateDataVector` was arrived at after a couple of iterations with @cjvogl. 

If I remember correctly, the idea was to have a future proof generic base class, viz. `ODEStateData`.
And have  `ODEStateDataVector` be a (currently only) concrete implementation of a `ODEStateData`.

I was trying to make a point to keep it _open_... and _general_...
But looking at it again, I guess you are right. Although it does do _assumptions_. 
As this function could be overridden in derived classes that is not necessarily an issue.

Reply:
I will try `state = Get(i)`  as standard implementation in the base class. I will test on my end.

Reply:
Not sure if this is helpful information, but the introduction of the `ODEStateData` interface was to separate definition from implementation (i.e., the abstract actions are defined in `ODEStateData` with the implementation of those actions, including choices on data storage, made in the derived classes).

Having `const Vector &Get` in the interface does impose a requirement that the derived classes implement the storage either as `Vector` objects or as something that can be cast to a `Vector` object; however, the [other comment](https://github.com/mfem/mfem/pull/3480/files/12927b36f0516a00c276fa54ce10ac1a5bb35ebb#r1549123727) gave the idea that, if desired down the road, that requirement could be removed by introducing a template parameter to the interface.

Regardless, @v-dobrev's point does make sense to me that having both `Get(int, Vector&)` and `const Vector& Get(int)` introduces some redundancy.

Reply:
Thank you @cjvogl for your insight. I'll try to digest the comments later today. But if I understand correctly you agree with the suggested made by @v-dobrev. I'll try to make  the suggested changes (only question being whether my interpretation of what you are saying is correct).

Reply:
> But if I understand correctly you agree with the suggested made by @v-dobrev.

Your understanding is correct!